addons/gamification/models/gamification_karma_tracking.py

```diff
@@ -1,69 +1,128 @@
+    reason = fields.Text(default=lambda self: _('Added Manually'), string='Description')
+    origin_ref = fields.Reference(
+        string='Source',
+        selection=lambda self: self._get_origin_selection_values(),
+        default=lambda self: f'res.users,{self.env.user.id}',
+    )
+    origin_ref_model_id = fields.Many2one(
+        'ir.model', 'Source Type', compute='_compute_origin_ref_model_id', store=True)
+
+    @api.depends('old_value', 'new_value')
+    def _compute_gain(self):
+        for karma in self:
+            karma.gain = karma.new_value - (karma.old_value or 0)
+
+    @api.depends('origin_ref')
+    def _compute_origin_ref_model_id(self):
+        for karma in self:
+            if not karma.origin_ref:
+                karma.origin_ref_model_id = False
+                continue
+
+            karma.origin_ref_model_id = self.env['ir.model']._get(karma.origin_ref._name).id
+
+    @api.model_create_multi
+    def create(self, values_list):
+        # fill missing old value with current user karma
+        for values in values_list:
+            if 'old_value' not in values and values.get('user_id'):
+                user = self.env['res.users'].browse(values['user_id'])
+                values['old_value'] = user.karma
+
+            if 'gain' in values and 'old_value' in values:
+                values['new_value'] = values['old_value'] + values['gain']
+                del values['gain']
+
+        return super().create(values_list)
 
     @api.model
-    def _consolidate_last_month(self):
-        """ Consolidate last month. Used by a cron to cleanup tracking records. """
-        previous_month_start = fields.Date.today() + relativedelta(months=-1, day=1)
-        return self._process_consolidate(previous_month_start)
-
-    def _process_consolidate(self, from_date):
-        """ Consolidate trackings into a single record for a given month, starting
-        at a from_date (included). End date is set to last day of current month
-        using a smart calendar.monthrange construction. """
-        end_date = from_date + relativedelta(day=calendar.monthrange(from_date.year, from_date.month)[1])
+    def _consolidate_cron(self):
+        """Consolidate the trackings 2 months ago. Used by a cron to cleanup tracking records."""
+        from_date = date_utils.start_of(fields.Datetime.today(), 'month') - relativedelta(months=2)
+        return self._process_consolidate(from_date)
+
+    def _process_consolidate(self, from_date, end_date=None):
+        """Consolidate the karma trackings, from the given date month.
+
+        The consolidation keeps, for each user, the oldest "old_value" and the most recent
+        "new_value", creates a new karma tracking with those values and removes all karma
+        trackings between those dates.
+        """
+        self.env['gamification.karma.tracking'].flush_model()
+
+        from_date = date_utils.start_of(from_date, 'month')
```

Comment:
I think I already said this, but we already receive the start of the month.

I think it would be better if this method just received dates and did not compute anything, it's strange to pass it a start_date and have it changed, the caller should compute what it needs.